File: src/main/java/sen/saloum/promeet/services/Impl/CandidatService.java
package sen.saloum.promeet.services.Impl;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import sen.saloum.promeet.dto.CandidatDTO;
import sen.saloum.promeet.dto.MacrosDTO;
import sen.saloum.promeet.mapstracts.CandidatMapper;
import sen.saloum.promeet.models.Candidat;
import sen.saloum.promeet.models.Macros;
import sen.saloum.promeet.repos.CandidatRepository;

import java.util.Optional;

@Service
public class CandidatService {

    private final CandidatRepository candidatRepository;
    private final CandidatMapper candidatMapper;

    public CandidatService(CandidatRepository candidatRepository, CandidatMapper candidatMapper) {
        this.candidatRepository = candidatRepository;
        this.candidatMapper = candidatMapper;
    }

    // Création d’un candidat
    public CandidatDTO createCandidat(CandidatDTO dto) {
        Candidat candidat = candidatMapper.toEntity(dto);
        return candidatMapper.toDTO(candidatRepository.save(candidat));
    }

    // Récupérer un candidat par ID
    public Optional<CandidatDTO> getCandidatById(Long id) {
        return candidatRepository.findById(id).map(candidatMapper::toDTO);
    }

    // Liste paginée de candidats
    public Page<CandidatDTO> getAllCandidats(Pageable pageable) {
        return candidatRepository.findAll(pageable)
                .map(candidatMapper::toDTO);
    }

    // Recherche par nom ou prénom (paginée)
    public Page<CandidatDTO> searchCandidats(String keyword, Pageable pageable) {
        return candidatRepository.searchByNomOrPrenom(keyword, pageable)
                .map(candidatMapper::toDTO);
    }
    public Optional<CandidatDTO> updateCandidat(Long id, CandidatDTO dto) {
        return candidatRepository.findById(id).map(existing -> {
            Candidat updated = candidatMapper.toEntity(dto);
            updated.setId(existing.getId()); // conserver l’ID
            return candidatMapper.toDTO(candidatRepository.save(updated));
        });
    }
    // Dans CandidatService
    public Optional<CandidatDTO> updateMacros(Long candidatId, MacrosDTO macrosDTO) {
        return candidatRepository.findById(candidatId).map(candidat -> {
            // On applique les changements des macros
            if (candidat.getMacros() == null) {
                candidat.setMacros(new Macros());
            }
            candidat.getMacros().setCv(macrosDTO.getCv());
            candidat.getMacros().setLm(macrosDTO.getLm());
            candidat.getMacros().setCvFibem(macrosDTO.getCvFibem());
            candidat.getMacros().setPriseDeNotes(macrosDTO.getPriseDeNotes());

            // Sauvegarder et renvoyer le DTO
            return candidatMapper.toDTO(candidatRepository.save(candidat));
        });
    }

    // Supprimer
    public void deleteCandidat(Long id) {
        candidatRepository.deleteById(id);
    }
}

File: src/app/services/candidat.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, of } from 'rxjs';
import { Candidat, Macros } from '../models/candidat';
import { Page } from '../models/page';

@Injectable({
  providedIn: 'root'
})
export class CandidatService {
  private apiUrl = 'http://localhost:8091/api/candidats'; // adapte selon ton backend

  constructor(private http: HttpClient) {}

  // Créer un candidat
  createCandidat(candidat: Candidat): Observable<Candidat> {
    return this.http.post<Candidat>(this.apiUrl, candidat);
  }

  // Récupérer tous les candidats (paginés)
  getCandidats(page: number, size: number): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}?page=${page}&size=${size}`);
  }

  // Récupérer un candidat par ID
  getCandidatById(id: number): Observable<Candidat> {
    return this.http.get<Candidat>(`${this.apiUrl}/${id}`);
  }

  // Mettre à jour un candidat complet
  updateCandidat(id: number, candidat: Candidat): Observable<Candidat> {
    return this.http.put<Candidat>(`${this.apiUrl}/${id}`, candidat);
  }

  // Supprimer un candidat
  deleteCandidat(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  // 🔥 Mettre à jour uniquement les macros
  updateMacros(id: number, macros: Macros): Observable<Candidat> {
    return this.http.put<Candidat>(`${this.apiUrl}/${id}/macros`, macros);
  }

  // Recherche par nom ou prénom
  searchCandidats(keyword: string, page: number, size: number): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}/search?keyword=${keyword}&page=${page}&size=${size}`);
  }
}
